sorts/heapsort.ts

```diff
@@ -0,0 +1,46 @@
+/* 
+Implementation of in-place heap sort.
+
+  complexity:
+    best time - O(N log N)
+    avg time - O(N log N)
+    worst time - O(N log N)
+
+    memory - O(1)
+    in-place - Yes
+    stable - No
+*/
+
+module.exports = function HeapSort(arr: number[]): number[] {
+  const left = (i: number): number => i * 2 + 1,
+    right = (i: number): number => i * 2 + 2;
+
+  function heapifyBottom(i: number, limit = arr.length): void {
+    while (1) {
+      // pretend if we try to acces to element below limit it's undefined
+      const l = left(i) < limit ? left(i) : arr.length,
+        r = right(i) < limit ? right(i) : arr.length;
+      if (arr[r] > arr[i] && arr[r] >= arr[l]) {
+        [arr[i], arr[r]] = [arr[r], arr[i]];
+        i = r;
+      } else if (arr[l] > arr[i] && (arr[l] >= arr[r] || !(r in arr))) {
+        [arr[i], arr[l]] = [arr[l], arr[i]];
```

Comment:
Minor: I'd move to `swap` function, since it's used in two places & for readability.